overlord/devicestate/users.go

```diff
@@ -0,0 +1,339 @@
+	}
+	if len(su.Series()) > 0 && !strutil.ListContains(su.Series(), series) {
+		return "", nil, fmt.Errorf(errorPrefix+"%q not in series %q", email, series, su.Series())
+	}
+	if len(su.Models()) > 0 && !strutil.ListContains(su.Models(), model) {
+		return "", nil, fmt.Errorf(errorPrefix+"%q not in models %q", model, su.Models())
+	}
+	if len(su.Serials()) > 0 {
+		if serialAs == nil {
+			return "", nil, fmt.Errorf(errorPrefix + "bound to serial assertion but device not yet registered")
+		}
+		serial := serialAs.Serial()
+		if !strutil.ListContains(su.Serials(), serial) {
+			return "", nil, fmt.Errorf(errorPrefix+"%q not in serials %q", serial, su.Serials())
+		}
+	}
+
+	if !su.ValidAt(time.Now()) {
+		return "", nil, fmt.Errorf(errorPrefix + "assertion not valid anymore")
+	}
+
+	gecos := fmt.Sprintf("%s,%s", email, su.Name())
+	opts := &osutil.AddUserOptions{
+		SSHKeys:             su.SSHKeys(),
+		Gecos:               gecos,
+		Password:            su.Password(),
+		ForcePasswordChange: su.ForcePasswordChange(),
+	}
+	return su.Username(), opts, nil
+}
+
+func (u *userManager) setupLocalUser(username, email string) error {
+	user, err := userLookup(username)
+	if err != nil {
+		return fmt.Errorf("cannot lookup user %q: %s", username, err)
+	}
+	uid, gid, err := osutil.UidGid(user)
+	if err != nil {
+		return err
+	}
+	authDataFn := filepath.Join(user.HomeDir, ".snap", "auth.json")
+	if err := osutil.MkdirAllChown(filepath.Dir(authDataFn), 0700, uid, gid); err != nil {
+		return err
+	}
+
+	// setup new user, local-only
+	if !u.safe {
+		u.state.Lock()
+	}
+	authUser, err := auth.NewUser(u.state, username, email, "", nil)
+	if !u.safe {
+		u.state.Unlock()
+	}
+	if err != nil {
+		return fmt.Errorf("cannot persist authentication details: %v", err)
+	}
+	// store macaroon auth, user's ID, email and username in auth.json in
+	// the new users home dir
+	outStr, err := json.Marshal(struct {
+		ID       int    `json:"id"`
+		Username string `json:"username"`
+		Email    string `json:"email"`
+		Macaroon string `json:"macaroon"`
+	}{
+		ID:       authUser.ID,
+		Username: authUser.Username,
+		Email:    authUser.Email,
+		Macaroon: authUser.Macaroon,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot marshal auth data: %s", err)
+	}
+	if err := osutil.AtomicWriteFileChown(authDataFn, []byte(outStr), 0600, 0, uid, gid); err != nil {
+		return fmt.Errorf("cannot write auth file %q: %s", authDataFn, err)
+	}
+
+	return nil
+}
+
+func (u *userManager) addUser(username string, email string, opts *osutil.AddUserOptions, createdUsers []UserResponse) ([]UserResponse, error) {
+
+	if err := osutilAddUser(username, opts); err != nil {
+		return nil, fmt.Errorf("cannot add user %q: %s", username, err)
+	}
+	if err := u.setupLocalUser(username, email); err != nil {
+		return nil, fmt.Errorf("%s", err)
+	}
+	createdUsers = append(createdUsers, UserResponse{
+		Username: username,
+		SSHKeys:  opts.SSHKeys,
+	})
+	return createdUsers, nil
```

Comment:
Changed as suggested. Idea was to avoid each caller called  `append`. But agree function signature is cleaner like this. Also as the function is to add a single user ( `addUser`) it's more natural it returns a single `UserResponse`.